gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/channel/WebSocketChannelizer.java

```diff
@@ -108,7 +109,7 @@ public void configure(final ChannelPipeline pipeline) {
                 closeOnProtocolViolation(false).allowExtensions(true).maxFramePayloadLength(settings.maxContentLength).build();
         pipeline.addLast(PIPELINE_REQUEST_HANDLER, new WebSocketServerProtocolHandler(GREMLIN_ENDPOINT,
                 null, false, false, 10000L, wsDecoderConfig));
-
+        pipeline.addLast("ws-user-agent-handler", new WsUserAgentHandler());
```

Comment:
Please create a JIRA to implement this for HTTP calls as well.

Reply:
This handler should go "after" authN has taken place. Else an attacker may perform a DoS attack by sending unauthenticated large string in user agent.

Reply:
Unfortunately I'm not sure this reordering is possible without substantial change. The user agent handler triggers on a HandShake.Complete event right as the handshake is finishing. The authentication and authorization handlers take effect after the handshake is finished. Since the user agent is sent as an http header though, it is limited by MaxHeaderSize which defaults to 8192. This should be sufficient protection against maliciously large headers being sent from unauthenticated users. I will add a test which confirms this behavior.

Reply:
I have made this JIRA for the HTTP implementation. https://issues.apache.org/jira/browse/TINKERPOP-2830

Reply:
> The authentication and authorization handlers take effect after the handshake is finished.

AFAIK, AuthN is done as part of the handshake itself. But irrespective of that, I think our new handler only executes when a handshake is complete. In which case our new handler's code will not execute for non authenticated requests, hence, there is no risk (contrary to what I mentioned earlier). Is my understanding right here?

Reply:
Unfortunately in my tests, the user agent handler always triggers before the authentication handler regardless of the order the handlers are added to the pipeline. This is the order of events as I understand it:

The handshake is completely controlled by Netty's `WebSocketServerProtocolHandler`. When the handshake is completed, that handler fires off a `WebSocketServerProtocolHandler.HandshakeComplete` event which contains the request headers with the user agent. This event is the only reasonable place I am aware of that the user agent can be extracted. After this event has fired, the handshake is complete and the connection has been successfully upgraded to a web socket connection. The authentication and authorization handlers then do their work by exchanging messages with the client through the open web socket connection. This cannot be done until after the web socket has been established.

I don't see any good way to delay the user agent handler from triggering until after authentication. We will need to rely on the maxHeaderSize config limiting the size of any incoming header. I am also working on a change which will impose a hard limit on the number of unique user agents being added to the metrics to protect against a malicious user trying to fill memory with millions of garbage user agents.